cmds/init/init.go

```diff
@@ -154,61 +163,71 @@ func main() {
-	cmdList := []string{
-		"/inito",
-		"/bbin/uinit",
-		"/buildbin/uinit",
-		"/bbin/sh",
-		"/bbin/rush",
-		"/buildbin/sh",
-		"/buildbin/rush",
-	}
+	osInitGo()
+	// If the os-specific runner failed, we can just keep going.
 
-	noCmdFound := true
 	for _, v := range cmdList {
-		if _, err := os.Stat(v); !os.IsNotExist(err) {
-
-			noCmdFound = false
-
-			initFlags := cmdline.GetInitFlagMap()
-			// systemd gets upset when it discovers it isn't really process 1
-			systemd, present := initFlags["systemd"]
-			systemdEnabled, boolErr := strconv.ParseBool(systemd)
-			if !present || boolErr != nil || systemdEnabled == false {
-				cmd = exec.Command(v)
-				cmd.Env = envs
-				cmd.Stdin = os.Stdin
-				cmd.Stderr = os.Stderr
-				cmd.Stdout = os.Stdout
-				if *test {
-					cmd.SysProcAttr = &syscall.SysProcAttr{Cloneflags: cloneFlags}
-				} else {
-					cmd.SysProcAttr = &syscall.SysProcAttr{Setctty: true, Setsid: true, Cloneflags: cloneFlags}
-				}
-				debug("Run %v", cmd)
-				if err := cmd.Run(); err != nil {
-					log.Print(err)
-				}
+		if _, err := os.Stat(v); os.IsNotExist(err) {
+			continue
+		}
+
+		// inito is (optionally) created by the u-root command when
+		// the u-root initramfs is merged with an existing initramfs that has
+		// a /init. The name inito means "original /init"
+		// There may be an inito if we are building on
+		// an existing initramfs. All initos need their
+		// own pid space.
+		var cloneFlags uintptr
+		if v == "/inito" {
+			cloneFlags = uintptr(syscall.CLONE_NEWPID)
+		}
+
+		cmdCount++
+		cmd = exec.Command(v)
+		cmd.Env = envs
+		cmd.Stdin, cmd.Stdout, cmd.Stderr = os.Stdin, os.Stdout, os.Stderr
+		if *test {
+			cmd.SysProcAttr = &syscall.SysProcAttr{Cloneflags: cloneFlags}
+		} else {
+			cmd.SysProcAttr = &syscall.SysProcAttr{Setctty: true, Setsid: true, Cloneflags: cloneFlags}
+		}
+		debug("Run %v", cmd)
+		if err := cmd.Start(); err != nil {
+			log.Printf("Error starting %v: %v", v, err)
+			continue
+		}
+		for {
+			var s syscall.WaitStatus
+			var r syscall.Rusage
+			if p, err := syscall.Wait4(-1, &s, 0, &r); p == cmd.Process.Pid {
+				debug("Shell exited, exit status %d", s.ExitStatus())
+				break
+			} else if p != -1 {
+				debug("Reaped PID %d, exit status %d", p, s.ExitStatus())
 			} else {
-				debug("Exec %v", v)
-				if err := syscall.Exec(v, []string{v}, envs); err != nil {
-					log.Print(err)
-				}
+				debug("Error from Wait4 for orphaned child: %v", err)
+				break
 			}
 		}
-		// only the first init needs its own PID space.
-		cloneFlags = 0
+		if err := cmd.Process.Release(); err != nil {
+			log.Printf("Error releasing %v:%v", v, err)
+		}
 	}
-
-	if noCmdFound {
+	if cmdCount == 0 {
 		log.Printf("init: No suitable executable found in %+v", cmdList)
 	}
 
+	// We need to reap all children before exiting.
+	log.Printf("init: Waiting for orphaned children")
+	for {
+		var s syscall.WaitStatus
+		var r syscall.Rusage
+		p, err := syscall.Wait4(-1, &s, 0, &r)
```

Comment:
Why not just use the Wait4 loop from above and keep looping? What's the point in having two of these?

Reply:
First loop is catching the child you spawned. Second loop is catching background children of the child you spawned. 

Consider the cast that we drop to a shell as the last thing.

% sleep 300 &
% exit

The first loop will catch the exit of the shell; then loop at the end of init will catch the exit of background processes. You should do both.